src/appleseed/renderer/kernel/lighting/bdpt/bdptlightingengine.cpp

```diff
@@ -83,37 +123,317 @@ namespace
+                    light_vertex.m_shading_basis,
+                    static_cast<Vector3f>(foundation::normalize(camera_vertex.m_position - light_vertex.m_position)),
+                    static_cast<Vector3f>(light_vertex.m_dir_to_prev_vertex),
+                    ScatteringMode::All,
+                    light_eval_bsdf);
+
+                Spectrum geometry = ComputeGeometryTerm(shading_context, shading_point, camera_vertex, light_vertex);
+                result = geometry * camera_eval_bsdf.m_beauty * light_eval_bsdf.m_beauty * camera_vertex.m_beta * light_vertex.m_beta;
+            }
+
+            // check if connection is near black or not
+            bool nearBlack = true;
+            for (size_t i = 0; i < result.size(); i++)
+            {
+                // need nearBlack() function for spectrum
+                if (result[i] > 1e-9)
+                {
+                    nearBlack = false;
+                    break;
+                }
+            }
+
+            if (nearBlack) { return; }
+
+            float numerator = ComputePathDensity(light_vertices, camera_vertices, s, t, s, t);
+            float denominator = 0.0;
+
+            // [p = 0, q = s + t], [p = 1, q = s + t - 1] ... [p = s + t - 2, q = 2]
+            for (int i = 0; i <= s + t - 2; i++)
+            {
+                denominator += ComputePathDensity(light_vertices, camera_vertices, s, t, i, s + t - i);
+            }
+
+            float miWeight = numerator / denominator;
+
+            assert(miWeight <= 1.0f);
+            radiance.m_beauty += miWeight * result;
+        }
+
         void compute_lighting(
             SamplingContext&        sampling_context,
             const PixelContext&     pixel_context,
             const ShadingContext&   shading_context,
             const ShadingPoint&     shading_point,
             ShadingComponents&      radiance) override      // output radiance, in W.sr^-1.m^-2
         {
-            PathVisitor light_path_visitor;
-            VolumeVisitor volume_visitor;
+            std::vector<BDPTVertex> camera_vertices;
+            std::vector<BDPTVertex> light_vertices;
 
-            PathTracer<PathVisitor, VolumeVisitor, true> light_path_tracer(     // true = adjoint
-                light_path_visitor,
-                volume_visitor,
-                ~0,
-                1,
-                ~0,
-                ~0,
-                ~0,
-                ~0,
-                shading_context.get_max_iterations());
+            // camera_vertices[0] is on shading_point
+            camera_vertices.reserve(num_max_vertices - 1); 
+
+            // light_vertices[0] is on light source
+            light_vertices.reserve(num_max_vertices);
+
+            trace_light(sampling_context, shading_context, &light_vertices);
+            trace_camera(sampling_context, shading_context, shading_point, &camera_vertices);
+
+            for (int s = 0;s < light_vertices.size() + 1;s++)
```

Comment:
Type mismatch, `s` must be of type `size_t`.

Basically `size_t` is the default integer type to use everywhere, unless you need signed integers in which case the default is `int`.